Evaluation docs

athitten opened the pull request:
> [!IMPORTANT]  
> The `Update branch` button must only be pressed in very rare occassions.
> An outdated branch is never blocking the merge of a PR.
> Please reach out to the automation team before pressing that button.

# What does this PR do ?

Adds docs for evaluation with [lm-eval-harness](https://github.com/EleutherAI/lm-evaluation-harness) integrated to NeMo, including guide to run it with NeMo 2.0 models.

**Collection**: [Note which collection this PR will affect]

# Changelog

- Add specific line by line info of high level changes in this PR.

# Usage

- You can potentially add a usage example below

```python
# Add a code snippet demonstrating how to use this 
```

# GitHub Actions CI

The Jenkins CI system has been replaced by GitHub Actions self-hosted runners.

The GitHub Actions CI will run automatically when the "Run CICD" label is added to the PR.
To re-run CI remove and add the label again.
To run CI on an untrusted fork, a NeMo user with write access must first click "Approve and run".

# Before your PR is "Ready for review"

**Pre checks**:

- [ ] Make sure you read and followed [Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md)
- [ ] Did you write any new necessary tests?
- [ ] Did you add or update any necessary documentation?
- [ ] Does the PR affect components that are optional to install? (Ex: Numba, Pynini, Apex etc)
  - [ ] Reviewer: Does the PR have correct import guards for all optional libraries?
  
**PR Type**:

- [ ] New Feature
- [ ] Bugfix
- [ ] Documentation

If you haven't finished some of the above items you can still open "Draft" PR.

## Who can review?

Anyone in the community is free to review the PR once the checks have passed.
[Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md) contains specific people who can review PRs to various areas.

# Additional Information

- Related to # (issue)

jgerh reviewed (commented):
Completed the technical doc review of docs/source/evaluation/evaluation_doc.rst and provided copyedits and suggested revisions for clarity and organization of content.

athitten commented on docs/source/evaluation/evaluation_doc.rst:
```diff
@@ -0,0 +1,83 @@
+
+    python scripts/llm/evaluation.py --nemo_checkpoint='/workspace/hf_llama3_8b_nemo2.nemo'
+
+Note: With the local executor run, it is required to manually kill 
```
Oh sorry my bad, had left the sentence incomplete here.

athitten commented on docs/source/evaluation/evaluation_doc.rst:
```diff
@@ -0,0 +1,101 @@
+
+.. note::
+
+Please refer to ``deploy`` and ``evaluate`` method in ``nemo/collections/llm/api.py`` to check all the argument
```
@jgerh could you please share what's the correct note syntax ? With this, I don't see it rendered anything different from normal text. Thanks!

athitten commented:
> Completed the technical doc review of docs/source/evaluation/evaluation_doc.rst and provided copyedits and suggested revisions for clarity and organization of content.

Thank you so much @jgerh! I have addressed all the comments and made changes.

Glorf commented on docs/source/evaluation/evaluation_doc.rst:
```diff
@@ -0,0 +1,101 @@
+    nemo_checkpoint = '/workspace/hf_llama3_8b_nemo2.nemo/'
+    api_endpoint = ApiEndpoint(nemo_checkpoint_path=nemo_checkpoint)
+    eval_target = EvaluationTarget(api_endpoint=api_endpoint)
+    eval_params = ConfigParams(top_p=1, temperature=1, top_k=1, limit_samples=2, num_fewshot=5)
```
Maybe worthwhile to briefly mention what these parameters mean? Or at least somewhat point to the dataclasses descriptions WDYT?

athitten replied:
Thanks @Glorf! Added pointer to the  dataclasses descriptions

Glorf commented on docs/source/evaluation/evaluation_doc.rst:
```diff
@@ -0,0 +1,101 @@
+.. code-block:: bash
+
+    python scripts/llm/evaluation.py --nemo_checkpoint='/workspace/hf_llama3_8b_nemo2.nemo' --slurm --nodes 1 
+    --devices 8 --container_image "nvcr.io/nvidia/nemo:dev" --tensor_parallelism_size 8
```
In the doc, should we maybe point to 25.02 or so after all?

athitten replied:
Added.

Glorf reviewed (approved):
Some small comments, overall LGTM!

jgerh commented on docs/source/evaluation/evaluation_doc.rst:
```diff
@@ -0,0 +1,101 @@
+
+.. note::
+
+When running locally with NeMo-Run, you will need to manually terminate the deploy process once evaluations are complete.
```
```suggestion
.. note::

   When running locally with NeMo-Run, you will need to manually terminate the deploy process once evaluations are complete.
```

athitten replied:
Had already made this change.

athitten commented on docs/source/evaluation/evaluation_doc.rst:
```diff
@@ -0,0 +1,101 @@
+Please refer to ``deploy`` and ``evaluate`` method in ``nemo/collections/llm/api.py`` to check all the argument
+options as these are just sample commands and don't share all arguments and their default settings.
+
+Run evaluations with NeMo-Run
```
Fixed.

athitten commented on docs/source/evaluation/evaluation_doc.rst:
```diff
@@ -0,0 +1,101 @@
+.. note::
+
+Please refer to ``deploy`` and ``evaluate`` method in ``nemo/collections/llm/api.py`` to check all the argument
+options as these are just sample commands and don't share all arguments and their default settings.
```
Had already made this change.

jgerh commented on docs/source/evaluation/evaluation_doc.rst:
```diff
@@ -0,0 +1,103 @@
+
+This guide provides detailed instructions on evaluating NeMo 2.0 checkpoints using the integrated `lm-evaluation-harness
+<https://github.com/EleutherAI/lm-evaluation-harness>`__ within the NeMo Framework. Supported benchmarks include
+``MMLU``, ``GSM8k``, ``lambada_openai``, ``winogrande``, ``arc_challenge``, ``arc_easy``, ``copa``.
```
``MMLU``, ``GSM8k``, ``lambada_openai``, ``winogrande``, ``arc_challenge``, ``arc_easy``, and ``copa``.

athitten replied:
Fixed.

jgerh commented on docs/source/evaluation/evaluation_doc.rst:
```diff
@@ -0,0 +1,103 @@
+.. note::
+
+  Please refer to ``deploy`` and ``evaluate`` method in ``nemo/collections/llm/api.py`` to check all the argument
+  options as these are just sample commands and don't share all arguments and their default settings. For more details
+  on arguments in ApiEndpoint and ConfigParams classes for evaluation refer to `nemo/collections/llm/evaluation/api.py
+  <https://github.com/NVIDIA/NeMo/blob/main/nemo/collections/llm/evaluation/api.py>`__.
```
For more details on arguments in the ApiEndpoint and ConfigParams classes for evaluation, refer to `nemo/collections/llm/evaluation/api.py
  <https://github.com/NVIDIA/NeMo/blob/main/nemo/collections/llm/evaluation/api.py>`__.

athitten replied:
Fixed.

jgerh reviewed (commented):
Completed a second review and added two comments

ko3n1g reviewed (approved):
docs only
